### PR TITLE
feat: Easy Event Deserialization 

### PR DESCRIPTION
**Pull request by jeromevdl**
**Issue #, if available:**

## Description of changes:

New feature in the serialization module: ability to easily extract the content from an Event (many built-in events).

Ex with API Gateway:

```java
import static software.amazon.lambda.powertools.utilities.EventDeserializer.from;

public class AppParams implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
    public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent event, Context context) {
        // will automatically extract the content from the body and serialize it as an Order
        Order order = extractDataFrom(event).as(Order.class); 

        // ...
    }
}
```

Ex with SQS:

```java
import static software.amazon.lambda.powertools.utilities.EventDeserializer.from;

public class AppParams implements RequestHandler<SQSEvent, String> {
    public String handleRequest(SQSEvent event, Context context) {
        // will automatically extract the content from records body, serialize them as Order, and return a list
        List<Order> orders = extractDataFrom(event).asListOf(Order.class); 

        // ...
    }
}
```

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-java/#tenets)
* [x] Update tests
* [ ] Update docs ==> _will update the doc once the code is reviewed & accepted_
* [x] PR title follows [conventional commit semantics]()

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static software.amazon.lambda.powertools.utilities.jmespath.Base64Function.decode;
+import static software.amazon.lambda.powertools.utilities.jmespath.Base64GZipFunction.decompress;
+
+public class EventDeserializer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EventDeserializer.class);
+
+    public static EventPart from(Object obj) {
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:33]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.

Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:33]: No real choice here, handling all built-in events

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            return new EventPart(event.getBody());
+        } else if (obj instanceof APIGatewayV2HTTPEvent) {
+            APIGatewayV2HTTPEvent event = (APIGatewayV2HTTPEvent) obj;
+            return new EventPart(event.getBody());
+        } else if (obj instanceof SNSEvent) {
+            SNSEvent event = (SNSEvent) obj;
+            return new EventPart(event.getRecords().get(0).getSNS().getMessage());
+        } else if (obj instanceof SQSEvent) {
+            SQSEvent event = (SQSEvent) obj;
+            return new EventPart(event.getRecords().stream().map(SQSEvent.SQSMessage::getBody).collect(Collectors.toList()));
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:49]: Please indent streaming in all if else branches, where applicable to have one operation per line. I know we have limited options to make if else prettier but atleast properly indenting things, make it much more readable.  for example

```suggestion
return new EventPart(event.getRecords().stream()
                    .map(SQSEvent.SQSMessage::getBody)
                    .collect(toList()));
```

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                if (contentMap != null) {
+                    return JsonConfig.get().getObjectMapper().convertValue(contentMap, clazz);
+                }
+                if (contentObject != null) {
+                    return (T) contentObject;
+                }
+                if (contentList != null) {
+                    throw new EventDeserializationException("The content of this event is a list, consider using 'extractDataAsListOf' instead");
+                }
+                throw new EventDeserializationException("Event content is null");
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:131]: Possible to validate this code path in tests? seems missing

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        public <T> List<T> asListOf(Class<T> clazz) {
+            if (contentList == null) {
+                throw new EventDeserializationException("Event content is null");
+            }
+            return contentList.stream().map(s -> {
+                try {
+                    return JsonConfig.get().getObjectMapper().reader().readValue(s, clazz);
+                } catch (IOException e) {
+                    throw new EventDeserializationException("Cannot load the event as " + clazz.getSimpleName(), e);
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:145]: Possible to validate this code path in tests? Seems missing

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                }
+                throw new EventDeserializationException("Event content is null");
+            } catch (IOException e) {
+                throw new EventDeserializationException("Cannot load the event as " + clazz.getSimpleName(), e);
+            }
+        }
+
+        public <T> List<T> asListOf(Class<T> clazz) {
+            if (contentList == null) {
+                throw new EventDeserializationException("Event content is null");
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:139]: Possible to validate this code path in tests? seems missing

File: powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java [hunk 6]
```diff
@@ -91,8 +91,14 @@ public Object around(ProceedingJoinPoint pjp,
                     event.getRecords().forEach(record -> validate(decode(record.getData()), inboundJsonSchema));
                 } else if (obj instanceof KafkaEvent) {
                     KafkaEvent event = (KafkaEvent) obj;
-                    event.getRecords().forEach((s, records) -> records.forEach(record -> validate(record.getValue(), inboundJsonSchema)));
-                }else if (obj instanceof KinesisAnalyticsFirehoseInputPreprocessingEvent) {
+                    event.getRecords().forEach((s, records) -> records.forEach(record -> validate(decode(record.getValue()), inboundJsonSchema)));
+                } else if (obj instanceof ActiveMQEvent) {
```
Comment (pankajagrawal16) [inline powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java:95]: I am correct in understanding that with this change, we are sort of covering this feature request too ? https://github.com/awslabs/aws-lambda-powertools-java/issues/179

File: powertools-validation/src/test/resources/sqs.json [hunk 7]
```diff
@@ -1,5 +1,5 @@
 {
-  "records": [
+  "Records": [
```
Comment (pankajagrawal16) [inline powertools-validation/src/test/resources/sqs.json:2]: Why this change ? 🤔

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        } else if (obj instanceof RabbitMQEvent) {
+            RabbitMQEvent event = (RabbitMQEvent) obj;
+            return new EventPart(event.getRmqMessagesByQueue().values().stream().flatMap(List::stream).map(r -> decode(r.getData())).collect(Collectors.toList()));
+        } else if (obj instanceof KinesisAnalyticsFirehoseInputPreprocessingEvent) {
+            KinesisAnalyticsFirehoseInputPreprocessingEvent event = (KinesisAnalyticsFirehoseInputPreprocessingEvent) obj;
+            return new EventPart(event.getRecords().stream().map(r -> decode(r.getData())).collect(Collectors.toList()));
+        } else if (obj instanceof KinesisAnalyticsStreamsInputPreprocessingEvent) {
+            KinesisAnalyticsStreamsInputPreprocessingEvent event = (KinesisAnalyticsStreamsInputPreprocessingEvent) obj;
+            return new EventPart(event.getRecords().stream().map(r -> decode(r.getData())).collect(Collectors.toList()));
+        } else {
```
[the remark this answers is not in the document]
Comment (pankajagrawal16) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:83]: I am thinking may be we can have a github action which somehow polls java lib weekly or daily or whatever and create an issue if for powertools if something new is added for events ?

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static software.amazon.lambda.powertools.utilities.jmespath.Base64Function.decode;
+import static software.amazon.lambda.powertools.utilities.jmespath.Base64GZipFunction.decompress;
+
+public class EventDeserializer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EventDeserializer.class);
+
+    public static EventPart extractDataFrom(Object obj) {
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:33]: Yep, I could have written the javadoc actually, will do it.

**Comment by msailes** [comment]
Would anything change post v4 of the [events library](https://github.com/aws/aws-lambda-java-libs/issues/282)? 

It might be worth a quick chat with @andclt.

File: powertools-serialization/pom.xml [hunk 10]
```diff
@@ -45,6 +45,14 @@
             <groupId>io.burt</groupId>
             <artifactId>jmespath-jackson</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-lambda-java-events</artifactId>
```
Comment (msailes) [inline powertools-serialization/pom.xml:50]: In v4 of the events lib, there will be a dependency on Jackson, I don't think that changes anything, just worth noting.

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/JsonConfig.java [hunk 11]
```diff
@@ -38,11 +36,7 @@ public static JsonConfig get() {
         return ConfigHolder.instance;
     }
 
-    private static final ThreadLocal<ObjectMapper> om = ThreadLocal.withInitial(() -> {
-        ObjectMapper objectMapper = new ObjectMapper();
-        objectMapper.configure(FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (msailes) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/JsonConfig.java:43]: Do we know why this was configured like this?

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:83]: Not all events make sense in here. I didn't add all, only those where there is a message or something meaningful to extract. Maybe I forgot some interesting ones.

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 2 above)
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:49]: OK

File: powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java (hunk 6 above)
Comment (jeromevdl) [reply powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java:95]: Yes, ActiveMQ and RabbitMQ have been added to the validation.

File: powertools-validation/src/test/resources/sqs.json (hunk 7 above)
Comment (jeromevdl) [reply powertools-validation/src/test/resources/sqs.json:2]: Records with big R is the correct one. But before the availability of the [aws-lambda-java-serialization](https://github.com/aws/aws-lambda-java-libs/tree/master/aws-lambda-java-serialization) library, we could not deserialize an SQS event properly (Mixins needed). This is why there was a small r. Now, I'm using the library in the test, so I could correct the json file.

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/JsonConfig.java (hunk 11 above)
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/JsonConfig.java:43]: It was configured like this for the tests in validation module, before using the java-lambda-serialization library. See discussion with Pankaj [here](https://github.com/awslabs/aws-lambda-powertools-java/pull/757#discussion_r809905638).

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:33]: javadoc ok

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 5 above)
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:139]: I changed a bit the logic here. This should not occur (hard to test). Adding a case and changed to IllegalStateException

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 4 above)
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:145]: Yes

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                    return JsonConfig.get().getObjectMapper().convertValue(contentMap, clazz);
+                }
+                if (contentObject != null) {
+                    return (T) contentObject;
+                }
+                if (contentList != null) {
+                    throw new EventDeserializationException("The content of this event is a list, consider using 'asListOf' instead");
+                }
+                // should not occur, except if the event is malformed (missing fields)
+                throw new IllegalStateException("Event content is null");
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:191]: It will be great to put a bit meaningful message for the users as well in case this happens at all, Like what could be the reason as mentioned in comments?

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+         * @param <T> type of object to return
+         * @return a list of objects of type T (deserialized from the content)
+         */
+        public <T> List<T> asListOf(Class<T> clazz) {
+            if (contentList == null) {
+                if (content != null || contentMap != null || contentObject != null) {
+                    throw new EventDeserializationException("The content of this event is not a list, consider using 'as' instead");
+                }
+                // should not occur, except if the event is really malformed
+                throw new IllegalStateException("Event content is null");
```
Comment (pankajagrawal16) [inline powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:209]: It will be great to put a bit meaningful message for the users as well in case this happens at all, Like what could be the reason as mentioned in comments?

**Review by pankajagrawal16 — COMMENTED** [review]
Minor suggestion, from code perspective, it looks good to me, so if you want to start documenting it.

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pankajagrawal16) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java:132]: Ok, may be i am overthinking this, we wait for doing anything more until we see it being a problem

File: powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java (hunk 13 above)
Comment (jeromevdl) [reply powertools-serialization/src/main/java/software/amazon/lambda/powertools/utilities/EventDeserializer.java]: done

**Comment by jeromevdl** [comment]
@pankajagrawal16 @msailes ready for further review / merge

**Review by msailes — APPROVED** [review]
👍🏻

File: docs/utilities/serialization.md [hunk 14]
```diff
@@ -3,7 +3,244 @@ title: Serialization Utilities
...
+
+### Key features
+
+* Easily deserialize the main content of an event (for example, the body of an API Gateway event)
+* 15+ built-in events (see the [list below](#built-in-events))
+
+### Getting started
+
+
+=== "Maven"
```
Comment (pankajagrawal16) [inline docs/utilities/serialization.md:18]: Please add gradle example too.

**Review by pankajagrawal16 — COMMENTED** [review]
Looks good. Just one minor suggestion on docs.

File: docs/utilities/serialization.md (hunk 14 above)
Comment (jeromevdl) [reply docs/utilities/serialization.md:17]: done

**Review by pankajagrawal16 — APPROVED** [review]
Great 🎉 ... cc @kozub
